syntax = "proto2";

package rv.protos;

import "rastervision/protos/raster_source.proto";

message ObjectDetectionGeoJSONFile {
    required string uri = 1;
}

message ClassificationGeoJSONFile {
    message Options {
        // The minimum IOA of a polygon and cell.
        optional float ioa_thresh = 1;

        // If use_intersection_over_cell is true, then use the area of the
        // cell as the denominator in the IOA. Otherwise, use the area of the
        // polygon.
        optional bool use_intersection_over_cell = 6;

        // If true, the class_id for a cell is the minimum class_id of the
        // boxes in that cell. Otherwise, pick the class_id of the box
        // covering the greatest area.
        optional bool pick_min_class_id = 2;

        // Optional class_id to use as the background class; ie. the one that
        // is used when a window contains no boxes. If not set, empty windows
        // have None set as their class_id.
        optional int32 background_class_id = 3;

        // Height and width of each cell (in pixels) in the spatial grid that
        // is laid over the raster.
        optional int32 cell_size = 4;

        optional bool infer_cells = 5 [default=false];
    }

    required string uri = 1;
    optional Options options = 2;
}

message SegmentationRasterFile {

    // An optional raster source from which the label data are to be
    // drawn.  If `src` is provided but `dst` is not, then label store
    // will be read-only.
    optional RasterSource src = 1;

    // An optional uri pointing to where the label data should be
    // stored.  If both `dst` is provided but `src` is not, then the
    // label store will be write-only.
    optional string dst = 2;

    // `src_classes` is a list of classes that appear in the `src`
    // RasterSource.  The strings contained in this array must be
    // recognizable by PIL ImageColor (e.g. '#ff7f00').  Given 24-bit
    // RGB input data, a string '#ff7f00' referes to pixels with whose
    // red channel is set to 0xff, whose green channel is set to 0x7f,
    // and whose blue channel is set to 0x00.
    //
    // `dst_classes` is a list of classes that appear in the asset
    // pointed-to by the `dst` uri string.  The integers contained in
    // this array must be representable as 8-bit unsigned values.
    //
    // There is a correspondence between `src_classes` and
    // `dst_classes`.  That correspondence is established by the
    // following logical procedure (though the actual implementation
    // may vary): dict(zip(src_classes, dst_classes)).  The
    // correspondence is used to map between source label values and
    // destination label values.
    repeated string src_classes = 3;
    repeated int32 dst_classes = 4 [packed=true];
}

message LabelStore {
    oneof label_store_type {
        ObjectDetectionGeoJSONFile object_detection_geojson_file = 1;
        ClassificationGeoJSONFile classification_geojson_file = 2;
        SegmentationRasterFile segmentation_raster_file = 3;
    }
}
